lte/gateway/python/magma/pipelined/ng_manager/node_state_manager.py

```diff
@@ -0,0 +1,179 @@
+        self._assoc_mon_thread = None
+
+        self.smf_assoc_version = 1
+        self.assoc_message_count = 0
+        self._assoc_established = False
+
+        self.smf_assoc_state = UPFAssociationState.STARTED
+        self._recovery_timestamp = Timestamp()
+
+        logging.info(" NGServicer : Node state manager launched ")
+
+    def _get_teid_pool_range(self):
+        #TEID_RANGE_INDICATION = 0 shows wild card as per 15.8, section 8.2.82
+        return self.TEID_RANGE_INDICATION, self.TEID_RANGE_VALUE
+
+    def _get_config(self, config_dict: Dict) -> NamedTuple:
+        def get_enodeb_if_ip(iface):
+            enode_if_ip = netifaces.ifaddresses(iface)
+            return enode_if_ip[netifaces.AF_INET][0]['addr']
+
+        def get_node_identifier(ng_params):
+            if ng_params and ng_params['node_identifier']:
+                return ng_params['node_identifier']
+            return get_enodeb_if_ip(config_dict['enodeb_iface'])
+
+        return self.GetLocalConfig(
+            downlink_ip=get_enodeb_if_ip(config_dict['enodeb_iface']),
+            node_identifier=get_node_identifier(config_dict['5G_feature_set'])
+        )
+
+    def _send_messsage_wrapper(self, node_message):
+        return send_node_state_association_request(node_message,\
+                                                   self._sessiond_setinterface)
+
+    def _send_association_request_message(self, assoc_message):
+
+        #Build the message
+        node_message=UPFNodeState(upf_id=self._node_id)
+        node_message.associaton_state.CopyFrom(assoc_message)
+
+        if self._send_messsage_wrapper(node_message) == True:
+            self.smf_assoc_state = assoc_message.assoc_state
+
+            if self.smf_assoc_state != UPFAssociationState.RELEASE:
+                self.smf_assoc_version += 1
+
+            self.assoc_message_count += 1
+            return True
+
+        return False
+
+    def send_association_setup_message(self):
+
+        teid_range_indicate, teid_range_value =\
+                  self._get_teid_pool_range()
+
+        #Create Node association setup message
+        assoc_message= \
+           UPFAssociationState(
+                           state_version=self.smf_assoc_version,
+                           assoc_state=UPFAssociationState.ESTABLISHED,
+                           feature_set=UPFFeatureSet(f_teid=True),
+                           recovery_time_stamp=self._recovery_timestamp.GetCurrentTime(),
+                           ip_resource_schema=\
+                             [UserPlaneIPResourceSchema(ipv4_address=self.config.downlink_ip,
+                                                        teid_range_indication=teid_range_indicate,
+                                                        teid_range=teid_range_value)]
+                           )
+
+        self._assoc_mon_thread = hub.spawn(self._monitor_association, assoc_message)
+
+    def _monitor_association(self, assoc_message, poll_interval=3):
```

Comment:
can you typehint thins, like `UPFAssociationState` and `int`

Reply:
Comments addresssed.